fop/fom_generic.c

```diff
@@ -433,12 +452,30 @@ M0_INTERNAL int m0_fom_tx_logged_wait(struct m0_fom *fom)
 {
 	struct m0_dtx   *dtx = &fom->fo_tx;
 	struct m0_be_tx *tx = m0_fom_tx(fom);
+	int              rc;
 
+	M0_ENTRY();
 	if (!fom_is_update(fom))
 		;
 	else if (dtx->tx_state == M0_DTX_DONE) {
-		if (m0_be_tx_state(tx) >= M0_BTS_LOGGED)
+		if (m0_be_tx_state(tx) >= M0_BTS_LOGGED) {
+			if (!fom->fo_local) {
+				rc = m0_fom_fdmi_record_post(fom);
+				/*
+				 * This reference is taken in fom_fol_rec_add().
+				 */
+				m0_be_tx_put(&fom->fo_tx.tx_betx);
+				/*
+				 * FDMI is reliable, so at this point there is
+				 * no way to handle the failure better than to
+				 * terminate the application.
+				 */
+				M0_ASSERT_INFO(rc == 0,
```

Comment:
I am worrying about these assertions here.
What can cause this failure?
What can we do in this case?
Can we skip some FDMI records?

Reply:
> What can cause this failure?

Already described in the comment above.

> What can we do in this case?

Already described in the comment above.

> Can we skip some FDMI records?

No, FDMI is reliable transport by design.

Reply:
I don't think we must panic and stop the Motr services here.
FDMI is designed to be reliable, but if something bad happens, we need to process this failure, and restart the post again.

Reply:
@huanghua78: please suggest a way to handle this error here.

Reply:
Motr itself should be "reliable" and "robust".
In FDMI source end, this post() is provide by plug-in from user.
If something wrong in a plug-in, we can report this error in IEM, and continue the service.

As an enterprise storage system, we should avoid panic the core software, especially in user provided code.

Reply:
BE doesn't have rollback, and FDMI is reliable by design. If FDMI service is not able to encode a data structure into a preallocated buffer than we have a bug in the code. In such case we can't rollback the transaction and we can't skip sending FDMI record, so there is no other option except to crash the process. We have `M0_ASSERT()`s about assumptions all over the code, and I'm adding one more.

> In FDMI source end, this post() is provide by plug-in from user.

No, it's not. FDMI source sends FDMI record over the network to FDMI plugin, and this is what is provided by the user. Please take a look at documentation in https://github.com/Seagate/cortx-motr/blob/main/fdmi/fdmi.c.

> If something wrong in a plug-in, we can report this error in IEM, and continue the service.

The error condition here is on FDMI source side, not on FDMI plugin side. There is no way I could think of to handle error here with IEM. If you think otherwise please elaborate.

If you provide a way to handle the error here I would check if it conforms to the design and I would be happy to implement it. If there is no such way than `M0_ASSERT()` is the best from whatever's left.

Reply:
>No, it's not. FDMI source sends FDMI record over the network to FDMI plugin, and this is what is provided by the user. Please take a look at documentation in https://github.com/Seagate/cortx-motr/blob/main/fdmi/fdmi.c.

As I asked before, what could cause the failure of this post()?
For example, if the BE tx may failure due to insufficient credit, then this is a bug.
If the network between the FDMI src and FDMI plug-in fails, is it possible to cause the failure?
So, in this case, this is not a bug. This is caused by some failure. We need to generate IEM alert messages, and so, this can be fixed.
If the FDMI plug-in side dies, is it possible to fail?


Or a more general question. If we are going to post a FDMI record from FDMI source to FDMI plug-in side, if there are some failures (network failures, or client error and not able to receive it), what should the Motr service do?
Panic and stop working?

Reply:
The existing implementation always returns 0 and the caller couldn't handle failure, so I made #700 to make this function `void`.

> As I asked before, what could cause the failure of this post()?

Nothing currently.

> If the network between the FDMI src and FDMI plug-in fails, is it possible to cause the failure?
Not in this case, the current implementation simply posts an FDMI record to a queue.

> If the FDMI plug-in side dies, is it possible to fail?

No, handling of this error is supposed to happen asyncrhonously, far from this call.

> Or a more general question. If we are going to post a FDMI record from FDMI source to FDMI plug-in side, if there are some failures (network failures, or client error and not able to receive it), what should the Motr service do?
Panic and stop working?

FDMI service is supposed to resend indefinitely, until the receiving end tells that this message could be discarded. Otherwise FDMI wouldn't be a reliable transport.